### ISSUE TITLE
runtime: frequent TestSegv failures since 2021-10-26

### ISSUE THREAD

**Issue by bcmills** [golang/go#49182]
@prattmic, possibly related to [CL 339989](https://golang.org/cl/339989) / #47522?

`greplogs --dashboard -md -l -e 'FAIL: TestSegv' --since=2021-10-01`

[2021-10-27T13:12:49-cfb5321/illumos-amd64](https://build.golang.org/log/f14faf19aeb3b952318d64f7f354d2bd062634ea)
[2021-10-27T08:50:27-bdefb77/illumos-amd64](https://build.golang.org/log/751a2dc072e6f4323fa154ad618003bf36b9b4fb)
[2021-10-27T08:50:27-bdefb77/solaris-amd64-oraclerel](https://build.golang.org/log/a93c55d3dbd2ece1cb886a89ffb58ecc3d2c8737)
[2021-10-27T05:33:58-ca5f65d/illumos-amd64](https://build.golang.org/log/9a99a3cae632593c7dd355f1460c87ba9222570b)
[2021-10-26T23:12:17-13eccaa/illumos-amd64](https://build.golang.org/log/590c00e19346fac110189428236ade6829c1e4ee)
[2021-10-26T23:12:17-13eccaa/solaris-amd64-oraclerel](https://build.golang.org/log/7e54025e89fab5cd936034ee744846db3260a0f0)
[2021-10-26T23:06:10-e5c5125/illumos-amd64](https://build.golang.org/log/085a9985cdbbdd3f8de4e16b07e83c6a4191d1af)
[2021-10-26T23:06:10-e5c5125/solaris-amd64-oraclerel](https://build.golang.org/log/965dc175927a82284a5e645ec50e683cc1f0b390)
[2021-10-26T22:05:53-80be4a4/illumos-amd64](https://build.golang.org/log/6465067b6d3ba6b114c4bd80cf99c3a442af4de0)
[2021-10-26T22:05:53-80be4a4/solaris-amd64-oraclerel](https://build.golang.org/log/1449cf9bd31d0ddcd0c724cffd2a4b279148d5b9)
[2021-10-26T21:32:57-86f6bf1/darwin-amd64-race](https://build.golang.org/log/77546c9a3ddde1e6b7982cb5dc204f9d9b963f47)
[2021-10-26T21:32:57-86f6bf1/illumos-amd64](https://build.golang.org/log/08f8315851bd0e20448544636d8df95ffccd15ca)

**Comment by bcmills** [comment]
`linux-arm-aws` is the only one that appears to still be flaky:

`greplogs --dashboard -md -l -e '(?ms)TestSegv.*unexpectedly saw "runtime: "'`

[2021-11-02T20:47:30-b246873/linux-arm-aws](https://build.golang.org/log/ff7f5f1b7c76631e0a9a7550eecce973e450619f)
[2021-11-02T17:01:14-62b29b0/linux-arm-aws](https://build.golang.org/log/f5461165c8693ca04c6c9f52a0ded03ab82c2205)
[2021-10-28T16:54:58-6bd0e7f/linux-arm-aws](https://build.golang.org/log/8f27de6d0af19898452531d7972e61df5d220a99)

**Comment by aclements** [comment]
This looks like the signal is landing while we're in the VDSO, presumably executing a kernel-provided atomic for casgstatus.

@prattmic, you looked at and fixed a few similar issues recently (e.g., 86f6bf18b013d570e89f57c2decaddca5ce2a847). Any insights on this one?

**Comment by aclements** [comment]
Yeah, the "unknown PC" is in the kernel-provided CAS implementation. The problem may just be that we're entirely missing the vdsoSP protection around the VDSO cas call (presumably the same thing applies to `memory_barrier<>`, too).

**Comment by prattmic** [comment]
Yes, I believe that is the case. I'll get those covered as well.

**Comment by jeremyfaller** [comment]
Mostly fixed. Not a Beta1 blocker as it's not a new breakage, just a stricter test.

**Comment by bcmills** [comment]
Looks like this is fixed on ARM and skipped on amd64, but still occasionally failing on some of the more exotic architectures.
(It's not clear to me whether the remaining failures are arch-specific bugs.)

`greplogs --dashboard -md -l -e 'FAIL: TestSegv ' --since=2021-11-13`

[2021-11-17T04:55:12-1d004fa/linux-mips64le-mengzhuo](https://build.golang.org/log/2ba30a59a0d9153badba232475f2fb813c85cbc7)
[2021-11-13T00:23:16-39bc666/linux-riscv64-unmatched](https://build.golang.org/log/61d916b32cca1bc547e9903e08eb129f4d1dc5a3)

**Comment by bcmills** [comment]
Looks like both of those failures are in the `SegvInCgo` subtest.

**Comment by bcmills** [comment]
https://storage.googleapis.com/go-build-log/3c4f5e79/linux-riscv64-jsing_db2bc678.log (A `linux-riscv64-jsing` SlowBot), also failed in `SegvInCgo`.

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e 'FAIL: TestSegv ' --since=2021-11-18`

[2021-12-02T20:45:47-3c6295d/netbsd-amd64-9_0-n2d](https://build.golang.org/log/d0a4092e2d398dbd3bb6f6cd1a691cd17adaad25)
[2021-12-02T16:48:07-36be0be/netbsd-amd64-9_0-n2d](https://build.golang.org/log/6002cf88bc515cf1c06700dbf4c8f104703c0c00)
[2021-11-24T21:11:45-b77f5f9/linux-mips64le-mengzhuo](https://build.golang.org/log/5a0560b2a80c156786c8e79fa45c4d58f17008b7)
[2021-11-24T20:59:14-67dd9ee/linux-s390x-ibm](https://build.golang.org/log/6b39bfc9414dbda03d8c7e668484920c5f65f763)
[2021-11-24T20:53:29-003e7fa/linux-riscv64-jsing](https://build.golang.org/log/d896e6ea3ecb2648aa93a876b18b4b40969f5481)
[2021-11-24T05:36:54-47db3bb/openbsd-amd64-70](https://build.golang.org/log/3168fa15ffa2fadd444946dbf78821a9463bf039)
[2021-11-23T19:31:05-0f64c21/linux-riscv64-unmatched](https://build.golang.org/log/91110be09739e3ccf069c4330fd2885f466f042e)
[2021-11-22T03:24:07-a287c4a/openbsd-amd64-70](https://build.golang.org/log/c89c8c322efdcf1e76243ea23f1af26a8145af7b)
[2021-11-18T02:17:23-ce7e501/linux-arm64-packet](https://build.golang.org/log/4e0568b45fa5f5531e1e13a7768f2d5bd8b4345d)

**Comment by prattmic** [comment]
The netbsd and openbsd failures there appear to be #49209.

**Comment by mknyszek** [comment]
Excluding the NetBSD and OpenBSD failures, all the ones that are left are in `TestSegvInCgo` like the two @bcmills posted about before. I *don't* think this is the same problem as the one @prattmic fixed earlier, because even an arm64 builder is failing there.

I'll poke at this one a bit.

**Comment by mknyszek** [comment]
In all these cases it looks like the signal is landing in some part of the cgocall path (not all that surprising), though I'm not sure I understand why gentraceback has issues producing a traceback in these cases.

**Comment by mknyszek** [comment]
oh oh, ok, actually for the linux/arm64 failure, the PC looks like the PC for the branch that @prattmic added in https://go.dev/cl/362977 (i.e. the top bits of the PC are `0xffff`). So there's some kind of other VDSO (?) call on at least this platform where there's more going on.

The other platforms' failures don't seem to look like this, however.

**Comment by prattmic** [comment]
The ARM failure could occur even with 363459479014bde19e83d9fb6781310f63fb0b45 if one of the kernel helpers made a CALL (thus setting LR). I thought I verified that none of the helpers we used did this, but I guess I should go back and double-check on the specific kernel version on the builders.

Beyond this, it looks like we have failures on linux-mips64le, linux-riscv64, and linux-s390x.

**Comment by mknyszek** [comment]
I'll note that 3634594 only applies to 32-bit ARM, AFAICT, so it shouldn't be taking that path. It could still be that we have some VDSO call that makes frames.

The RISC-V failures are kind of interesting because there's no hexdump of the stack -- I'm not really sure why. Notably, on `cgoSigtramp`, the LR is 0 (`cgoSigtramp` is a thin wrapper around `sigtramp` that just jumps and links a zero LR.

**Comment by prattmic** [comment]
[2021-12-08T23:00:01-f5ddd94/linux-386](https://build.golang.org/log/b483007e2f3c3ff2e9cd5b80401e523633ca09f8)
[2021-12-04T04:41:31-cd5f2cf/linux-mips64le-mengzhuo](https://build.golang.org/log/e5ba4aec35956e74cd1f22e784cefee2d5734971)

**Comment by prattmic** [comment]
In the [linux-386](https://build.golang.org/log/b483007e2f3c3ff2e9cd5b80401e523633ca09f8) failure, `PC=0xf7f3db09` seems likely to be in the VDSO.

The VDSO load address is randomized, but this seems in approximately the correct range. Additionally, page offset `0xb09` is a valid instruction just after a syscall, which seems like a likely place to get a signal delivered.

```
   0xf7fd1b00 <__kernel_vsyscall>:      push   %ecx
   0xf7fd1b01 <__kernel_vsyscall+1>:    push   %edx
   0xf7fd1b02 <__kernel_vsyscall+2>:    push   %ebp
   0xf7fd1b03 <__kernel_vsyscall+3>:    mov    %esp,%ebp
   0xf7fd1b05 <__kernel_vsyscall+5>:    sysenter 
   0xf7fd1b07 <__kernel_vsyscall+7>:    int    $0x80
   0xf7fd1b09 <__kernel_vsyscall+9>:    pop    %ebp
   0xf7fd1b0a <__kernel_vsyscall+10>:   pop    %edx
   0xf7fd1b0b <__kernel_vsyscall+11>:   pop    %ecx
   0xf7fd1b0c <__kernel_vsyscall+12>:   ret  
```

Go doesn't use `__kernel_vsyscall` directly, but libc does. Maybe we get there somehow in a context where the signal handler doesn't realize this is cgo?

**Comment by prattmic** [comment]
e.g., `__kernel_vsyscall` is reachable from [`cgo_start_thread`](https://cs.opensource.google/go/go/+/master:src/runtime/proc.go;l=2255;drc=9b0de0854d5a5655890ef0b2b9052da2541182a3), which is called via `asmcgocall`.

It looks like the signal handler won't properly detect raw `asmcgocall` calls as cgo calls because `sighandler` checks [`g.m.incgo`](https://cs.opensource.google/go/go/+/master:src/runtime/signal_unix.go;l=690;drc=master), which is only set in `cgocall`.

@cherrymui does this sound correct to you, or did I miss some detection mechanism?

**Comment by cherrymui** [comment]
@prattmic I think that is right. Direct `asmcgocall` is used in limited situations, but I guess it is still possible to receive an external signal.

VDSO calls made from Go would have m.vdsoPC and m.vdsoSP set, so we can tell. But if it is made from non-Go code then we can't tell.

**Comment by cherrymui** [comment]
I still think it is too strong to require that we can unwind the stack any time we receive an external signal. It think it is important to print SIGSEGV, but the unwinding could be best effort. Maybe we could try unwinding the curg stack if the signal is landed on g0 (even if incgo is not set) so it could be somewhat helpful.

**Comment by prattmic** [comment]
> I still think it is too strong to require that we can unwind the stack any time we receive an external signal.

I tend to agree that as a general feature getting correct tracebacks when an external SIGSEGV lands in an esoteric location is not important to users. I've been pushing on these bugs primarily to support testing. To ensure that the common cases work and don't regress, I want to be able to write a straightforward test that tracebacks work. It is very difficult to write such a test if there are lots of edge cases that don't work properly.

For now, I'll skip this part of the test on 386, and I've filed #50504 as a low priority issue for this `asmcgocall` problem.

**Comment by bcmills** [comment]
The `linux-mips64le-mengzhuo` failures are still occurring on the builder. It's not obvious to me whether that has the same root cause as the `386` failure mode, so I filed  it separately as #50605.

**Comment by bcmills** [comment]
@golang/runtime, can this be re-triaged? The test still has a `testenv.SkipFlaky` call on `darwin`, `illumos`, and `solaris` referring to this issue — if there is some other tracking issue for the test (particularly given that `darwin` is a first-class platform), then the skip should be updated to refer to that issue.

**Comment by mknyszek** [comment]
Hello from triage. :) I think we didn't get to this because our incoming queue was quite busy. @prattmic will take a look at breaking it up. @cherrymui is working on a rewrite of the test that will split out the failures a bit better.

**Comment by cherrymui** [comment]
There are various TestSegv issues that watchflakes are tracking, e.g. #59443. I think we can close this as a dup. If this happens again watchflakes will post on or reopen an existing issue, or open a new one.

**Comment by bcmills** [comment]
@cherrymui, this issue is more about the followup work needed to diagnose the existing skips, particularly on `darwin` and on `linux/386` (which are both first-class ports):
https://cs.opensource.google/go/go/+/master:src/runtime/crash_cgo_test.go;l=667-675;drc=6dd3bfbed6f17e7789f092e96408c00c227a8b68

**Comment by bcmills** [comment]
That is: relying on `watchflakes` to report this issue is not appropriate because the relevant failures are already being skipped.

**Comment by cherrymui** [comment]
I think we know we are not always able to traceback from asynchronous interrupts in libc calls. I'm not sure if there is anything we can do. Or you mean we should push hard to make that work?

The linux/386 VDSO case may be fixable (or already fixed).

**Comment by cherrymui** [comment]
Sounds good. Thanks.
